codice/ddf
# DDF-5180 Create S3 Content Storage Provider

#### What does this PR do?
Create a Storage Provider for writing/reading products into/from S3 buckets.

#### Who is reviewing it? 
@codice/core-apis 
@jhunzik 
@sgalla 
@harrison-tarr

#### Ask 2 committers to review/merge the PR and tag them here.
@beyelerb 
@ryeats

#### How should this be tested?
**Configure the S3 Storage Provider:**
1. Start up a local DDF .
2. Start the "catalog-core-3storageprovider" feature.
3. Stop the "catalog-core-localstorageprovider" bundle.
4. Configure the "Content S3 Storage Provider" feature to include your S3 implmentation account credentials (access key/secret key) and to point to the correct S3 endpoint/region/bucket. 

**Test the Create operation:**
1. Upload a file through the Search UI, verify in S3 that the product was written to your configured S3 bucket.

**Test the Read operation:**
1. After uploading a product, In the Search UI, click on the link to download the resource, and verify that you were able to successfully download the product.

**Test the Update operation:**
1. Using a REST client (eg Postman), configure the HTTP URL as: https://localhost:8993/services/catalog/{metacardId}
2. Set the HTTP method as PUT to indicate an update.
3. Click on the "Body" tab and click "form-data".
4. Select a file to update the existing product with, see screenshot below:
![Screen Shot 2019-08-07 at 4 33 05 PM](https://user-images.githubusercontent.com/5935039/62664624-1fd37a80-b931-11e9-9663-daacd936b79c.png)
5. Send the request, query the Catalog via the Search UI and verify the metacard updated.
6. Verify in S3 that the old product updated to the new product.
7. Verify when trying to download the resource, the new product gets downloaded.

**Test the Delete operation**
1. In the DDF Karaf Console, type "catalog:removeall -f", to remove all catalog entries.
2. Verify in S3 all products are deleted.

#### Checklist:

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.
